crates/ruma-events-macros/src/event_enum.rs

```diff
@@ -1071,3 +1071,40 @@ impl EventEnumEntry {
         Ok(EventEnumVariant { attrs, ident })
     }
 }
+
+pub(crate) fn expand_from_impls_derived(input: DeriveInput) -> TokenStream {
+    let variants = match &input.data {
+        Data::Enum(DataEnum { enum_token: _, brace_token: _, variants }) => variants,
+        _ => panic!("this derive macro only works with enums"),
+    };
+
+    let from_impls = &variants
+        .iter()
+        .map(|variant| match &variant.fields {
+            syn::Fields::Unnamed(fields) => {
+                if fields.unnamed.len() == 1 {
+                    let inner_struct =
+                        &fields.unnamed.first().unwrap().ty.clone().into_token_stream();
+                    let var_ident = &variant.ident.to_token_stream();
+                    let id = &input.ident;
+                    quote! {
+                        impl From<#inner_struct> for #id {
```

Comment:
Can we instead implement `From<T> for #id where T: Into<#inner_struct>`? That would allow both `AnySyncMessageEvent` => `AnySyncRoomEvent` (using the blanket `From<T> for T` impl from `std` to fulfill the bound) and `SyncMessageEvent<MessageEventContent>` => `AnySyncRoomEvent`.

Reply:
I'm not sure we can implement `From<T> for #id where T: Into<#inner_struct>`.

When I tried, I got a bunch of 
```
conflicting implementations of trait `std::convert::From<_>` for type `enums::AnyRoomEvent`
first implementation here (.....)
```

I'm guessing that the compiler is being conservative and assumes that someone could implement `From<OneVariant> for AnotherVariant`, causing an overlap in the impls?

Reply:
Oh, right. Somebody in an external crate could implement `Into<AnySyncMessageEvent>` as well as `Into<AnySyncStateEvent>` for the same type. Hmm. In that case let's leave this as-is, it doesn't hurt though we should have another issue for allowing more conversions here in one way or another.